File: tests/test_beam_layout.py
"""S89: unit tests for drawing/beam_layout.py (cutout-aware beam + posts).

Judges the fix for B10 / B4-B6 without any rendering (vision-independent):
  1. FLAT deck (no cutout) reproduces the legacy post layout EXACTLY -- the
     no-zone permit set must stay byte-identical.
  2. DEEP front notch: no post lands inside the notch (the B10 bug), beam steps.
  3. SHALLOW front notch: a single straight beam absorbs it as a cantilever
     (option b) and still strands no post.
  4. No segment ever overhangs past the cantilever limit.
"""
import sys
from pathlib import Path

BACKEND = Path(__file__).resolve().parents[1] / "backend"
sys.path.insert(0, str(BACKEND))

from drawing.beam_layout import (  # noqa: E402
    compute_beam_layout, front_edge_profile, _legacy_posts,
)

FAILS = []


def check(name, cond, detail=""):
    tag = "OK  " if cond else "FAIL"
    if not cond:
        FAILS.append(name)
    print(f"  [{tag}] {name}" + (f"  {detail}" if detail else ""))


def _cut(x, y, w, d):
    return {"rect": {"x": x, "y": y, "w": w, "d": d}}


def post_in_any_cut(post_xy, cut_rects):
    """True if a post center sits inside any cutout rect (= stranded)."""
    for (px, py) in post_xy:
        for cr in cut_rects:
            r = cr["rect"]
            if r["x"] <= px <= r["x"] + r["w"] and r["y"] <= py <= r["y"] + r["d"]:
                return (px, py)
    return None


# --- 1. FLAT deck: byte-identical to legacy ---------------------------------
for (W, D, NP) in [(20, 12, 3), (16, 12, 3), (32, 14, 4), (10, 10, 2), (40, 16, 5)]:
    lay = compute_beam_layout(W, D, [], NP, cantilever_max=3.0)
    legacy = _legacy_posts(W, NP)
    got = lay["segments"][0]["posts"]
    check(f"flat {W}x{D} np={NP}: legacy posts preserved", got == legacy,
          f"{got} vs {legacy}")
    check(f"flat {W}x{D}: single beam at depth-1.5",
          not lay["stepped"] and abs(lay["segments"][0]["beam_y"] - (D - 1.5)) < 1e-6,
          f"beam_y={lay['segments'][0]['beam_y']}")

# --- 2. DEEP front notch (the B10 repro: 20x14, notch x[6,14] to y=8) --------
deep_cuts = [_cut(6, 8, 8, 6)]  # front cutout, edge pulled to y=8 (cd=6)
prof = front_edge_profile(20, 14, deep_cuts)
check("deep notch: profile has 3 levels 14/8/14",
      [round(e[2]) for e in prof] == [14, 8, 14], str(prof))
deep = compute_beam_layout(20, 14, deep_cuts, 3, cantilever_max=3.125)
strand = post_in_any_cut(deep["post_xy"], deep_cuts)
check("deep notch: NO post stranded in the notch", strand is None,
      f"stranded {strand}" if strand else "clean")
check("deep notch: beam steps to follow the edge", deep["stepped"] is True)
check("deep notch: a post sits under the shallow strip (beam_y ~6.5)",
      any(abs(py - 6.5) < 1e-6 for (_px, py) in deep["post_xy"]),
      str(deep["post_xy"]))

# --- 3. SHALLOW front notch (cd=1: absorbed by cantilever, option b) ---------
shallow_cuts = [_cut(6, 13, 8, 1)]  # edge pulled only to y=13 (cd=1)
shallow = compute_beam_layout(20, 14, shallow_cuts, 3, cantilever_max=3.125)
check("shallow notch: single straight beam (no step)", shallow["stepped"] is False)
strand2 = post_in_any_cut(shallow["post_xy"], shallow_cuts)
check("shallow notch: NO post stranded", strand2 is None,
      f"stranded {strand2}" if strand2 else "clean")

# --- 4. Cantilever limit respected in every case ----------------------------
for name, cuts in [("deep", deep_cuts), ("shallow", shallow_cuts)]:
    lay = compute_beam_layout(20, 14, cuts, 3, cantilever_max=3.125)
    worst = max(s["max_cant"] for s in lay["segments"])
    check(f"{name}: no overhang over cantilever_max (worst={worst})",
          not lay["over_limit"])

print()
if FAILS:
    print(f"BEAM LAYOUT: {len(FAILS)} FAILURE(S): {FAILS}")
    sys.exit(1)
print("BEAM LAYOUT: all checks passed")

File: tests/test_beam_layout.js
// ============================================================
// S101 -- JS BEAM LAYOUT GUARD
//
// Locks the JS port of backend/drawing/beam_layout.py (computeBeamLayout,
// _legacyPosts, _postsForSegment). Before the port the frontend computed a
// flat legacy post row from width alone, so a notched deck showed 3 posts on
// screen where the PDF correctly showed 6, and a post could sit stranded over
// the cutout void.
//
// The CRITICAL invariant here is the flat-deck one: with no front cutout the
// layout must reproduce _legacyPosts EXACTLY. Plain decks are the overwhelming
// majority of the product's output and their permit set must not move.
//
// Python-side equivalence is asserted separately, and more strongly, by
// tests/test_frontend_parity.py, which runs both engines on the same configs
// and compares materials line-for-line.
//
// Run: node tests/test_beam_layout.js
// ============================================================
global.window = global;
const fs = require('fs');
const path = require('path');
const ROOT = path.join(__dirname, '..');
eval(fs.readFileSync(path.join(ROOT, 'backend/static/js/stairGeometry.js'), 'utf8'));

// NOTE: `global.window = global` means stairGeometry.js's own top-level
// function declarations are already globals here, so binding them to consts of
// the SAME name is a redeclaration error. Use distinct local names.
const blCompute = window.computeBeamLayout;
const blLegacy = window._legacyPosts;
const blSegment = window._postsForSegment;

let failures = 0;
function check(cond, msg) {
  if (!cond) { console.log('  [FAIL] ' + msg); failures++; }
}
function eqArr(a, b) {
  return a.length === b.length && a.every((v, i) => Math.abs(v - b[i]) < 1e-9);
}

console.log('\nS101 JS beam layout');

// ---- 1. Flat deck reproduces the legacy row byte-for-byte ----
console.log('1. Flat deck == legacy positions (the no-regression invariant)');
[[20, 14, 3], [40, 12, 5], [10, 10, 2], [32, 16, 4], [13, 12, 3]].forEach(function (cfg) {
  const W = cfg[0], D = cfg[1], nP = cfg[2];
  const legacy = blLegacy(W, nP);
  const bl = blCompute(W, D, [], nP, 2.0, 1.5, 8.0);
  const got = bl.postXY.map(function (xy) { return xy[0]; });
  check(eqArr(got, legacy), 'flat ' + W + 'x' + D + ' nP=' + nP +
        ' expected ' + JSON.stringify(legacy) + ' got ' + JSON.stringify(got));
  check(bl.stepped === false, 'flat ' + W + 'x' + D + ' must not be stepped');
  check(bl.overLimit === false, 'flat ' + W + 'x' + D + ' must not be over limit');
  check(Math.abs(bl.segments[0].beamY - (D - 1.5)) < 1e-9,
        'flat ' + W + 'x' + D + ' beamY must be depth-1.5');
});

// A single post centers, matching the Python num_posts == 1 branch.
check(eqArr(blLegacy(20, 1), [10]), 'single post centers at width/2');

// ---- 2. Front cutout forces a real change ----
console.log('2. Front-reaching cutout changes the layout');
const cut = [{ rect: { x: 8, y: 10, w: 4, d: 4 } }];  // reaches front on D=14
const flat14 = blCompute(20, 14, [], 3, 2.0, 1.5, 8.0);
const notched = blCompute(20, 14, cut, 3, 2.0, 1.5, 8.0);
const flatX = flat14.postXY.map(function (p) { return p[0]; });
const notchX = notched.postXY.map(function (p) { return p[0]; });
check(!eqArr(flatX, notchX), 'a front cutout must change post positions');
check(notchX.length >= flatX.length,
      'notched deck should not have FEWER posts than flat (got ' +
      notchX.length + ' vs ' + flatX.length + ')');

// ---- 3. No post is stranded over the cutout void ----
// A post inside the cutout footprint must sit BEHIND it (beamY <= cutout y),
// never out at the original front edge where there is no deck above it.
console.log('3. No post stranded over the void');
notched.postXY.forEach(function (xy) {
  const x = xy[0], y = xy[1];
  const insideX = x > 8 - 1e-9 && x < 12 + 1e-9;
  if (insideX) {
    check(y <= 10 + 1e-9,
          'post at x=' + x + ' is inside the cutout span but sits at y=' + y +
          ' (front edge) -- stranded over the void');
  }
});

// ---- 4. A non-front-reaching cutout leaves the front edge alone ----
console.log('4. Interior cutout does not move the front beam');
const interior = [{ rect: { x: 8, y: 4, w: 4, d: 2 } }];  // y+d = 6, well short of 14
const interiorBL = blCompute(20, 14, interior, 3, 2.0, 1.5, 8.0);
check(eqArr(interiorBL.postXY.map(function (p) { return p[0]; }), blLegacy(20, 3)),
      'a cutout that does not reach the front edge must leave posts unchanged');
check(interiorBL.stepped === false, 'interior cutout must not step the beam');

// ---- 5. _postsForSegment: the S90 short-segment guard ----
// With the 2ft end inset the end posts sit (segW-4) apart, so a ~4ft segment
// would stack two posts at nearly the same x. Below MIN_POST_SEP it must
// collapse to ONE centered post.
console.log('5. Short-segment guard (S90 coincident-post bug)');
check(blSegment(0, 4, 8).length === 1, '4ft segment must yield 1 post');
check(Math.abs(blSegment(0, 4, 8)[0] - 2) < 1e-9, '4ft segment post centers at 2');
check(blSegment(0, 5, 8).length === 1, '5ft segment must yield 1 post');
check(blSegment(0, 6, 8).length === 2, '6ft segment must yield 2 distinct posts');
check(eqArr(blSegment(0, 6, 8), [2, 4]), '6ft segment posts at [2,4] (unchanged)');
blSegment(0, 20, 8).forEach(function (x, i, arr) {
  if (i > 0) check(arr[i] - arr[i - 1] > 1e-9, 'posts must be strictly increasing');
});

// ---- 6. max_beam_span is respected ----
console.log('6. Post spacing never exceeds max_beam_span');
[4, 6, 8, 10].forEach(function (maxSpan) {
  const posts = blSegment(0, 40, maxSpan);
  for (let i = 1; i < posts.length; i++) {
    check(posts[i] - posts[i - 1] <= maxSpan + 1e-6,
          'span ' + (posts[i] - posts[i - 1]).toFixed(2) +
          ' exceeds max ' + maxSpan);
  }
});

// ---- 7. Empty / degenerate input is safe ----
console.log('7. Degenerate input');
check(blCompute(20, 14, null, 3, 2.0, 1.5, 8.0).postXY.length === 3,
      'null cutRects must behave as flat');
check(blCompute(20, 14, undefined, 3, 2.0, 1.5, 8.0).stepped === false,
      'undefined cutRects must behave as flat');
const defaulted = blCompute(20, 14, [], 3, 2.0);
check(eqArr(defaulted.postXY.map(function (p) { return p[0]; }), blLegacy(20, 3)),
      'omitted setback/maxBeamSpan must fall back to the Python defaults');

console.log('');
if (failures === 0) {
  console.log('JS BEAM LAYOUT: all checks passed');
  process.exit(0);
} else {
  console.log('JS BEAM LAYOUT: ' + failures + ' check(s) FAILED');
  process.exit(1);
}
